### ISSUE TITLE
No IPv6 support in uTP/DHT

### ISSUE THREAD

**Issue by vuori** [mafintosh/discovery-swarm#45]
I was trying out Dat and noticed that IPv6 is not used to reach peers. This pretty seriously limits its usefulness for me, so I looked around the internals a bit. On a high level, I think the following changes would be needed in `discovery-swarm` to fix this:

  * Create separete IPv4 and IPv6 uTP sockets.
  * Run separate instances of `bittorrent-dht` for IPv4 and IPv6 on each socket (technically these could be merged, but makes the implementation a lot more complex for relatively little benefit).
  * Hide the distinction from higher layers (`discovery-channel` and up). It seems that this would happen pretty much automagically due to the event architecture.

On the lower levels, looks like this needs:

  * Support for binding IPv6 sockets in `utp-native`. Seems like a minor change, `libutp` already has support so just needs selecting the right `libuv` functions based on the input address.
  * Support for BIP-32 in `bittorrent-dht` and the `k-rpc` modules. A bit of work, but BIP-32 is pretty simple, and with instances handling a single address family this shouldn't need any major restructuring.
  * `dns-discovery`: no idea, not going to go there for now, would continue using the current (IPv4) socket as before.

I may have some time to work on this in the near future. Does the approach sound reasonable?

**Comment by vuori** [comment]
If I understand the workings of the DNS discovery system correctly, it relies on the server recording the transmitting IP of the host that 's announcing a service. That would require communicating with the `datprotocol.com` DNS servers over IPv6, but none of them have AAAA records. So it looks like the infrastructure would need to be fixed first before client-side changes can go anywhere?

**Comment by pfrazee** [comment]
@vuori that's correct, but less daunting than it sounds. We'd just need to update the dns-discovery module to support ipv6. @mafintosh can deploy the new server relatively quickly

**Comment by DaniellMesquita** [comment]
Reference: #1084 (duplicate)

**Comment by DaniellMesquita** [comment]
@vuori , how Dat could support i2p?

**Comment by vuori** [comment]
I'm not familiar with i2p. If you mean IPv6, it turned out that the rabbit hole went a bit too deep for a weekend project, so the issue has been on the shelf.

**Comment by RangerMauve** [comment]
@vuori I'm going to be doing some stability work on hyperswarm, I'll see if I can put in IPv6 support in my spare time.

**Comment by DaniellMesquita** [comment]
@RangerMauve how Dat could support Tor or i2p?

**Comment by RangerMauve** [comment]
@DaniellMesquita Dat can work across any transport. What you'd need to do is find some sort of peer discovery mechanism for Tor or i2p (Maybe a DHT exists for this?), open a connection to a peer, and use `hyperdrive.replicate()` to create a replication stream and pipe it into the peer.
